[clang][NFC] Annotate `Type` bit-fields with `clang::preferred_type`

Endilll opened the pull request:
This patch adds `clang::preferred_type` annotations to Type-related bit-fields. At the moment only debug info takes advantage of this annotation. See more in https://github.com/llvm/llvm-project/pull/69104

This patch also propagates underlying type of several enums from bit-field declaration to enum declaration. I don't see how having them diverge helps.

Endilll committed:
[clang][NFC] Annotate `Type` bit-fields with `clang::preferred_type`

Endilll committed:
Run clang-format

Endilll committed:
Revert "Run clang-format"

This reverts commit 21689b56d1fc1db0b2263e8049ff656d3757ad36.

Endilll commented:
Formatting is good. I created a PR against `clang-format` to convince the tool: https://github.com/llvm/llvm-project/pull/70360

cor3ntin commented on clang/include/clang/Basic/Linkage.h:
```diff
@@ -20,7 +20,7 @@ namespace clang {
 /// Describes the different kinds of linkage
 /// (C++ [basic.link], C99 6.2.2) that an entity may have.
-enum Linkage : unsigned char {
+enum Linkage : unsigned {
```
can you explain this change?

Endilll replied:
This case is similar to `TypeDependence` we discuss in another thread, except that I can't find references to this type in data member declaration in our headers, which suggests that bit-field it stored in has hard-coded width, and has nothing to do with underlying type declared here.

cor3ntin commented on clang/include/clang/AST/DependenceFlags.h:
```diff
@@ -49,7 +49,7 @@ struct ExprDependenceScope {
 
 struct TypeDependenceScope {
-  enum TypeDependence : uint8_t {
+  enum TypeDependence : unsigned {
```
Can you explain this change? Making Type bigger than it is can have a pretty large impact on the ressource need of clang.

Endilll replied:
The only place this enum is stored is bit-field I annotate in this patch. Like it was with `TypeAlignment`, I'm only encoding status quo that actual underlying type for this enum is unsigned. There is no impact on memory footprint, because I'm yet to see us deriving bit-field width from underlying type of enum.

Endilll replied:
In this particular case, width of bit-field is not hardcoded, but derived from enumerators via `llvm::BitWidth<TypeDependence>`

cor3ntin replied:
I had a chat with @Endilll offline.
The reason for this change is that `preferred_type` (#69104) warns when the enumerator has a different underlying type as the bit field. https://godbolt.org/z/zz99s6M3j

This warning does seem a bit off to me - but I really like its intent.
The warning only triggers on enumerators (which probably make sense as a general warning may have too many false positives), but I think the warning is too restrictive.
If the underlying type has the same signedness and has a size equal to **or smaller**  than the the size of the type of the bitfield, I don't think a warning should be produced.

I have convinced myself that the changes here are most likely fine (`TypeDependence` is (afaict) only stored in TypeBitfields, as far as i can tell, and this uses the number of bits of the greatest enumerator, which i did not initially realize).

Yet I do not like the idea that we would change the underlying type of all enums used in bitfields for the sake of that warning.

My preferred solution would be to make `-Wbitfield-type` less aggressive and not change the underlying types of enums is this patch.

cor3ntin replied:
@erichkeane @AaronBallman wdyt?

erichkeane replied:
>>My preferred solution would be to make -Wbitfield-type less aggressive and not change the underlying types of enums is this patch.

Yeah, this I think is my preference.

AaronBallman replied:
Now that we're reviewing https://github.com/llvm/llvm-project/pull/70632 should we revert the changes to underlying types from this patch?

Endilll replied:
Yeah, I'm putting this on hold for a bit, and refactor enumerations in NFC commits, so that this PR lands without "holes" because some enums are not complete at the point where I put `preferred_type`. After that underlying type changes this PR does shouldn't be necessary anymore.

Endilll replied:
I reverted changes to underlying types. This should be good to go now.

Endilll commented:
@tbaederr As highlighted in https://github.com/llvm/llvm-project/pull/69104, this attribute doesn't change semantics of the program, so it's safe to ignore.

tbaederr commented:
Sure but clang 16 will emit a warning for them: https://godbolt.org/z/jnsc4336G - are we passing `-Wno-unknown-attributes` to the build?

Endilll commented:
That's a good point. I'm not opposed to wrap this attribute if we must. I'd like to hear from @AaronBallman on this matter.

erichkeane commented:
We definitely need to wrap these in a macro/hide them in a macro somehow.  This will fail `-Werror` on GCC/MSVC/Older Clang compilers.

AaronBallman commented:
> We definitely need to wrap these in a macro/hide them in a macro somehow. This will fail `-Werror` on GCC/MSVC/Older Clang compilers.

+1, the macro would live in https://github.com/llvm/llvm-project/blob/main/llvm/include/llvm/Support/Compiler.h along with the other attribute macros.

Endilll committed:
Wrap attribute in a macro

Endilll committed:
Merge remote-tracking branch 'upstream/main' into annotate-type-with-preferred-type

Endilll committed:
Add missing annotations

This is made possible by several NFC commits that made some enums complete at the point of declaration of bit-fields.

Endilll committed:
Revert changes to underlying type of enums

Endilll committed:
Merge remote-tracking branch 'upstream/main' into annotate-type-with-preferred-type

AaronBallman commented on llvm/include/llvm/Support/Compiler.h:
```diff
@@ -569,4 +569,12 @@ void AnnotateIgnoreWritesEnd(const char *file, int line);
 
+/// \macro LLVM_PREFERRED_TYPE
+/// Adjust type of bit-field in debug info.
+#if __has_attribute(preferred_type)
```
```suggestion
#if LLVM_HAS_CPP_ATTRIBUTE(clang::preferred_type)
```
or leave the condition as-is and change the expansion to using `__attribute__(())` instead. Just making sure they both match up. (If we're using this from C interfaces, then we probably want to use the GNU-style spelling.)

AaronBallman reviewed (approved):
LGTM aside from a nit with the new macro (feel free to fix when landing).

Endilll committed:
Use GNU attribute syntax for `preferred_type`
